src/pytorch_lightning/loops/epoch/training_epoch_loop.py

```diff
@@ -292,9 +292,8 @@ def on_advance_end(self) -> None:
 
         # if training finished, defer exit to the parent. this assumes there will be enough time in between
         # which might not be the case depending on what's in the `*_epoch_end` hooks
-        if not self._is_training_done:
-            # if fault tolerant is enabled and process has been notified, exit.
-            self.trainer._exit_gracefully_on_signal()
+        if not self._is_training_done and self.trainer.received_sigterm:
```

Comment:
This logic is not equivalent to what we had before. Correct me if I'm wrong, but this looks like it will be (much?) slower than master because we are making an reduce collective call on every training step. This isn't the case in master, because in the previous logic we were testing whether fault tolerant training was enabled through `if _fault_tolerant_training()`.

Reply:
You are correct. The purpose of this PR is to untie this signal handling logic from fault tolerance, so this is expected (not meant to be equivalent).

These checks are run at the epoch end.

I don't have any good suggestion here. Would you suggest that we don't all_reduce?

Reply:
> These checks are run at the epoch end.

The check runs with the condition `not self._is_training_done` which is True in most training steps. So it's not just at the end of the epoch.

> Would you suggest that we don't all_reduce?

Need to think about this. I want to ask @tchaton. In theory, every process should get the signal regardless of where the process was created. But I'm no sure. We should be able to verify this.

Reply:
Based on this blog: https://the-fonz.gitlab.io/posts/python-multiprocessing/ I think we don't need the reduction as each spawned process will register the signal handler.

Reply:
From what I remember, it was required as SIGTERM won't always be passed or they could be race conditions. I don't think it harms to have it.
I would let you be the judge.